python/paddle/fluid/dygraph/dygraph_to_static/list_transformer.py

```diff
@@ -0,0 +1,169 @@
+# limitations under the License.
+
+from __future__ import print_function
+
+import gast
+import astor
+from paddle.fluid.dygraph.dygraph_to_static.static_analysis import AstNodeWrapper
+from .static_analysis import AstNodeWrapper, NodeVarType, StaticAnalysisVisitor
+
+
+class ListTransformer(gast.NodeTransformer):
+    """
+    This class transforms python list used in control flow into Static Graph Ast
+    """
+
+    def __init__(self, wrapper_root):
+        assert isinstance(
+            wrapper_root, AstNodeWrapper
+        ), "Input non-AstNodeWrapper node for the initialization of ListTransformer."
+        self.wrapper_root = wrapper_root
+        self.root = wrapper_root.node
+        self.name_of_list_set = set()
+        self.list_name_to_updated = dict()
+
+        self.static_analysis_visitor = StaticAnalysisVisitor(self.root)
+        self.node_to_wrapper_map = self.static_analysis_visitor.get_node_to_wrapper_map(
+        )
+        var_env = self.static_analysis_visitor.get_var_env()
+        var_env.cur_scope = var_env.cur_scope.sub_scopes[0]
+        self.scope_var_type_dict = var_env.get_scope_var_type()
+
+    def transform(self):
+        self.visit(self.root)
+        self.replace_list_with_tensor_array(self.root)
+
+    def visit_Assign(self, node):
+        if self._update_list_name_to_updated(node):
+            return node
+        return node
+
+    def visit_If(self, node):
+        self.generic_visit(node)
+        self._transform_list_append_in_control_flow(node)
+        return node
+
+    def visit_While(self, node):
+        self.generic_visit(node)
+        self._transform_list_append_in_control_flow(node)
+        return node
+
+    def visit_For(self, node):
+        self.generic_visit(node)
+        self._transform_list_append_in_control_flow(node)
+        return node
+
+    def _transform_list_append_in_control_flow(self, node):
+        for child_node in gast.walk(node):
+            if self._need_to_array_write_node(child_node):
+                child_node.value = \
+                    self._to_array_write_node(child_node.value)
+
+    def _need_to_array_write_node(self, node):
+        if isinstance(node, gast.Expr):
+            if isinstance(node.value, gast.Call):
+                if self._is_list_append_tensor(node.value):
+                    return True
+
+        return False
+
+    def replace_list_with_tensor_array(self, node):
+        for child_node in gast.walk(node):
+            if isinstance(child_node, gast.Assign):
+                if self._need_to_create_tensor_array(child_node):
+                    child_node.value = self._create_tensor_array()
+
+    def _is_list_append_tensor(self, node):
+        """
+        a.append(b): a is list, b is Tensor
+        """
+        assert isinstance(node, gast.Call)
+        # 1. is .append
+        if not isinstance(node.func, gast.Attribute):
+            return False
+        if node.func.attr != 'append':
+            return False
```

Comment:
But if `node` is `self.a.append(b)`, the function `_is_list_append_tensor ` will return `false` because the AST is `Call -> Attribute -> Attribute` which is not as expected.

More cases should be considered or add a TODO here.

Reply:
As we discussed offline, `self.a.append(b)` has been taken into account, so the code can handle `self.a.append(b)`. 
if the code is `self.a.append(b)`, node.func.attr is 'append'.